File: my-folder/0496-next-greater-element-i/solution.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> tempStack = new Stack<>();
        Map<Integer,Integer> greaterMap = new HashMap<>();
        for(int i=nums2.length-1;i>=0;i--){
            while(!tempStack.isEmpty() && nums2[i]>=tempStack.peek()) tempStack.pop();
            if(tempStack.isEmpty()) greaterMap.put(nums2[i],-1);
            else greaterMap.put(nums2[i],tempStack.peek());
            tempStack.push(nums2[i]);
        }
        int[] ans = new int[nums1.length];
        for(int i=0;i<nums1.length;i++){
            ans[i]=greaterMap.get(nums1[i]);
        }
        return ans;
    }
}

File: my-folder/0496-next-greater-element-i/solution.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> st;
        vector<int> ans(nums1.size(),-1);
        unordered_map<int,int> mp;
        for(int i = nums2.size()-1;i>=0;i--){
            while(!st.empty() && nums2[i]>=st.top()) st.pop();
            if(!st.empty()) mp[nums2[i]]=st.top();
            st.push(nums2[i]);
        }
        for(int i = 0;i<nums1.size();i++){
            if(mp.find(nums1[i])!=mp.end()) ans[i]=mp[nums1[i]];
            else ans[i]=-1;
        }
        return ans;
        
    }
};
